### ISSUE TITLE
Origin-Agent-Cluster

### ISSUE THREAD

**Issue by Malvoz** [helmetjs/helmet#275]
The [`Origin-Isolation`](https://html.spec.whatwg.org/multipage/origin.html#origin-isolation) header provides a mechanism to allow web applications to isolate their origins.

The [WICG explainer](https://github.com/WICG/origin-isolation) talks about the relation with [COEP](https://github.com/helmetjs/helmet/issues/198)/[COOP](https://github.com/helmetjs/helmet/issues/177)/[CORP](https://github.com/helmetjs/helmet/issues/176):

> Deploying `COOP` + `COEP` can be a significant burden for sites, e.g. in terms of how it requires all of their embedded content to be updated with `CORP`, `COEP`, and `COOP` headers.

> this separate header would be much easier to deploy than `COOP` + `COEP`, since it does not require any embeddee cooperation.

<hr>

Implementation status:

- https://chromestatus.com/feature/5683766104162304
- https://bugzilla.mozilla.org/show_bug.cgi?id=1665474
- https://bugs.webkit.org/show_bug.cgi?id=216618

**Comment by EvanHahn** [comment]
Thanks for reporting. It seems like this header is in early days, but probably worth supporting eventually.

**Comment by EvanHahn** [comment]
It looks like it's now been renamed to `Origin-Agent-Cluster`, per [the spec][0]. It looks like it was [added in Chrome 88][1] but is in no other major browsers yet.

If someone would like to make a pull request to add this header, I'd love it! Here's some guidance:

- We can't enable this header by default in Helmet 4, as it'd be a breaking change. It'll need to be disabled by default.
- Here's how you'd use this header by itself:

  ```js
  // Sets "Origin-Agent-Cluster: ?1"
  app.use(helmet.originAgentCluster());
  ```
- Here's how you'd use the header as part of the "greater" Helmet middleware:

  ```js
  app.use(helmet({
    originAgentCluster: true,
  }));
  ```

  [We currently check that no values are `true`][2]. We'll either need to make an exception for `originAgentCluster` or allow it for all middlewares; I'd only allow it for `originAgentCluster` for now.
- Setting the header to `undefined` or `false` should have no effect:

  ```js
  // All three of these should be equivalent:
  helmet();
  helmet({ originAgentCluster: false });
  helmet({ originAgentCluster: undefined });
  ```
- Unlike the other middlewares, I would _not_ expect this to be published as a separate npm package until someone requests it.
- Create a new source file, `middlewares/origin-agent-cluster/index.ts`. It'll look similar to `middlewares/x-download-options/index.ts`.
- Create a new test file, `test/origin-agent-cluster.test.ts`. It should look similar to `test/x-download-options.test.ts`.
- Add new tests in `test/index.ts` to make sure we can use it from the top level.
- Update `index.ts` to include everything necessary for `originAgentCluster`.
- Update the changelog (`CHANGELOG.md`).
- Add the built files to `"files"` in `package.json`.

Let me know if I can help at all.

[0]: https://whatpr.org/html/6214/origin.html#origin-keyed-agent-clusters
[1]: https://chromestatus.com/feature/5683766104162304
[2]: https://github.com/helmetjs/helmet/blob/08b5f4644ee5e8d6488fd169ce5d5c2da0766281/index.ts#L57-L61

**Comment by EvanHahn** [comment]
This was added in #286 and released in Helmet v4.4.0. Closing.

**Comment by timdown** [comment]
I'm not sure enabling this is appropriate as a default for Helmet. It has subtle implications and trade-offs and is not intended as a security feature (yeah, I know you somewhat disagree with that part). The following line from [this pretty authoritative page](https://web.dev/origin-agent-cluster/) sticks out:

> You shouldn't just roll out origin-keying and hope for the best.
